### ISSUE TITLE
Epic: A2C trainer + benchmark harness (learning curves + throughput)

### ISSUE THREAD

**Issue by rjwalters** [rjwalters/thrust#150]
## Background

Thrust's algorithm portfolio after the SAC epic (#136) is:
- **PPO** (on-policy, clipped surrogate, single + multi-agent)
- **DQN** (off-policy, discrete)
- **SAC** (off-policy, continuous control)
- **PSRO / NFSP** (multi-agent meta-solvers)

**Gap**: ROADMAP Milestone 6 also calls out **A2C** (Advantage Actor-Critic, Mnih et al. 2016, "Asynchronous Methods for Deep RL", arXiv:1602.01783 — the synchronous A2C variant). A2C is the simplest on-policy actor-critic: it is the *single-update-per-rollout, no-clipping* special case of the PPO family. It is high value as (a) a pedagogically clean baseline, (b) a fast-converging CartPole reference, and (c) the natural counterpart for the **benchmark harness** this epic also lands: learning curves AND throughput/wall-clock numbers, which the repo currently has no infrastructure for (`criterion` is a dep but the `[[bench]]` block is commented out and there is no `benches/` dir).

This is `/architect` work because it makes a reuse-vs-fork decision against the PPO trainer, and it stands up a brand-new benchmark harness (criterion + a learning-curve emit format) that did not exist before.

### 1. A2C algorithm surface — sibling `src/train/a2c/` module, REUSE everything else

**Decision: a new sibling `src/train/a2c/` module (`config.rs`, `loss.rs`, `trainer.rs`, `stats.rs` re-using PPO's `TrainingStats`), NOT a parameterization of `PPOTrainerBurn`.** Rationale (lower risk):
- `PPOTrainerBurn::train_step` is a proven, tested, bit-stable path that PSRO/NFSP/single-agent PPO all depend on. Threading an "A2C mode" flag through its clipped-surrogate / minibatch-shuffle / KL-early-stop / entropy-collapse logic would mean touching the most load-bearing trainer in the repo and risking the multi-agent callers. A sibling module keeps that blast radius at zero.
- A2C's loop is genuinely *simpler* (no `n_epochs`, no minibatching, no clip, no KL early-stop), so a flag-parameterized PPO would carry dead branches; a clean ~150-LOC `A2cTrainer` is easier to read and audit.

**What is REUSED (no new code):**
- `MlpBurnPolicy` (`src/policy/mlp.rs`) — same actor-critic net; A2C calls `evaluate_actions(obs, actions) -> (log_probs, entropy, values)` and `get_action_host`, exactly like PPO. The `MlpBurnConfig { seed: Some(..) }` seeded path (PR #137) threads reproducibility.
- `BurnOptimizer` / `BackendOptimizer` (`src/train/optimizer.rs`) — same move-through `Option<P>` ownership model and `.step(lr, module, grads)` pattern as `PPOTrainerBurn`.
- **GAE** — the example-level `compute_gae` helper (and `src/buffer/rollout/gae.rs`) computes advantages + returns identically; A2C uses the SAME GAE path (A2C is "GAE-λ + one update").
- `CartPole` env (`src/env/games/cartpole.rs`) + `EnvPool` rollout-collection pattern from `examples/games/cartpole/train_cartpole_modern.rs`.

**Where A2C DIVERGES from PPO (the new ~150 LOC in `a2c/`):**
- **Loss** (`a2c/loss.rs`): policy-gradient loss `-mean(log_prob * advantage_detached)` (NO ratio, NO clip), value loss `0.5 * MSE(V(s), returns)` (NO value-clip), entropy bonus `-ent_coef * mean(entropy)`. `total = pg_loss + value_coef * value_loss + ent_coef * entropy_loss`.
- **Loop** (`a2c/trainer.rs`): ONE gradient step per rollout — no epoch loop, no minibatch shuffle, no KL early-stop. Optional global-norm grad clip via `max_grad_norm`. Optional advantage normalization (config knob, default on, matching PPO). Keep PPO's entropy-collapse guard (it is generic).
- **Signature**: `A2cTrainer::train_step(observations, actions, advantages, returns, evaluate_fn) -> A2cStats`. Note it does NOT take `old_log_probs`/`old_values` — A2C is on-policy with no importance ratio, so those are unused.

### 2. `A2cConfig` knobs + classic defaults

Mirrors `DQNConfig`/`SacConfig`/`PPOConfig` builder + `validate()` + seeded-determinism conventions.

| Knob | Default | Notes |
|---|---|---|
| `learning_rate` | `7e-4` | classic A2C (RMSProp/Adam) value, Mnih 2016 / SB3 A2C default |
| `gamma` | `0.99` | discount |
| `gae_lambda` | `1.0` | classic A2C uses full n-step returns (λ=1); GAE-λ generalizes it |
| `value_coef` | `0.5` | value loss weight |
| `entropy_coef` | `0.01` | entropy bonus |
| `n_steps` | `5` | classic A2C rollout length (Mnih 2016 uses 5-step returns) |
| `num_envs` | `16` | synchronous parallel actors |
| `max_grad_norm` | `0.5` | global-norm clip |
| `normalize_advantages` | `true` | zero-mean/unit-var per rollout (matches PPO) |
| `seed` | `0` | bit-exact reproducibility (threads `MlpBurnConfig::with_seed`) |

`validate()`: `learning_rate > 0`, `0 <= gamma <= 1`, `0 <= gae_lambda <= 1`, `value_coef >= 0`, `entropy_coef >= 0`, `n_steps > 0`, `num_envs > 0`, `max_grad_norm > 0`.

### 3. CartPole benchmark target + always-on smoke bar

- **Env**: `CartPole` (discrete, 4-dim obs, 2 actions) — the natural A2C benchmark.
- **"Solved" convention**: mean reward >= 195 over 100 consecutive episodes (standard CartPole-v0 bar). A2C with the defaults above reaches this within roughly 100-200k env steps.
- **Always-on fast unit test** (`a2c/trainer.rs` `#[cfg(test)]`, no feature gate beyond `training`): construct `A2cTrainer` on `Autodiff<NdArray<f32>>`, run ONE `train_step` against a tiny synthetic batch, assert (a) `total_steps() == 1`, (b) all stat fields (`policy_loss`, `value_loss`, `entropy`, `total_loss`) are finite. This is the matching-pennies-analogue smoke test — fast, deterministic, runs in CI on every PR.
- **Slow convergence test** (`tests/test_a2c_cartpole.rs`, gated on `training`, NOT in the fast path): seeded A2C on CartPole reaches mean episode length >= 195 over the final 100 episodes within a fixed step budget. Marked `#[ignore]` or kept out of the default CI fast lane so it does not slow PRs, exactly like the SAC Pendulum convergence test.

### 4. Learning-curve benchmark (reward vs env-steps, A2C-vs-PPO, reproducible)

- The A2C example (`examples/games/cartpole/train_cartpole_a2c.rs`) writes a **CSV** `(env_steps, mean_episode_reward)` to a path given by an env var (e.g. `CURVE_CSV=/tmp/a2c_cartpole.csv`), one row per logging interval, seeded via `A2cConfig::seed`.
- The existing PPO example gets the SAME opt-in CSV emit (additive, behind the same env var) so the two curves are directly comparable on the same env/seed/step-budget.
- Reproducibility: both runs fix `seed`, `num_envs`, `n_steps`, total step budget; the CSV is the artifact a human (or a follow-on plotting script) overlays. No plotting code is added in-tree — emitting the data is the deliverable.

### 5. Throughput benchmark (criterion `benches/`)

- Create `benches/trainer_throughput.rs` (criterion, `harness = false`).
- Benchmarks: (a) `a2c_cartpole_steps_per_sec` — env-step + rollout-collect + one A2C update, reported as elements/sec via `Throughput::Elements(n_steps * num_envs)`; (b) `a2c_train_step` — time of a single `train_step` on a fixed synthetic batch; (c) the SAME two for PPO (`ppo_train_step`, `ppo_update`) for apples-to-apples comparison. All on `Autodiff<NdArray<f32>>` (CPU), seeded.
- Cargo wiring: UN-comment and adjust the `[[bench]]` block:
```toml
[[bench]]
name = "trainer_throughput"
harness = false
required-features = ["training"]
```
criterion is already in `[dev-dependencies]`.

### 6. Decomposition into Builder-tractable PRs (DAG)

Four PRs. Dependency DAG: **A -> B -> {C, D}**.

- **PR A — `A2cConfig`** (`src/train/a2c/config.rs` + `mod.rs` skeleton + `train/mod.rs` re-export). Blast radius: additive new module; one re-export line in `train/mod.rs`. No deps. *Ready to build.*
- **PR B — `A2cTrainer` + `a2c/loss.rs` + always-on smoke test** (`src/train/a2c/loss.rs`, `src/train/a2c/trainer.rs`). Reuses `MlpBurnPolicy` + `BurnOptimizer`. Includes the item-3 fast unit test. Blast radius: additive new module. **Depends on A** (needs `A2cConfig`).
- **PR C — CartPole A2C example + learning-curve CSV + slow convergence test** (`examples/games/cartpole/train_cartpole_a2c.rs`, CSV emit added to the PPO example too, `tests/test_a2c_cartpole.rs`, `[[example]]` block in Cargo.toml). Blast radius: new example + one test + additive example-block; touches the PPO example only to add an opt-in CSV writer behind an env var. **Depends on B** (needs a working `A2cTrainer`).
- **PR D — throughput benchmark harness** (`benches/trainer_throughput.rs`, un-comment `[[bench]]` in Cargo.toml). Blast radius: new `benches/` dir + Cargo bench block. **Depends on B** (benchmarks the `A2cTrainer`; PPO benches reuse existing infra).

```
#A ──> #B ──┬──> #C  (example + learning curves + convergence test)
            └──> #D  (criterion throughput harness)
```

C and D are independent of each other and can build in parallel once B merges.

---
Child implementation issues filed per item 6; numbers + DAG in a follow-up comment. This issue stays open as the tracking epic.

**Comment by rjwalters** [comment]
## Child implementation issues filed

Decomposition from item 6 is now on GitHub. Dependency DAG: **A -> B -> {C, D}**.

| PR | Issue | Title | Label | Depends on |
|---|---|---|---|---|
| A | #151 | A2cConfig (hyperparameters + builder + validate) | `loom:issue` (ready) | — |
| B | #152 | A2cTrainer + a2c/loss.rs + always-on smoke test | `loom:blocked` | #151 |
| C | #153 | CartPole example + learning-curve CSV + convergence test | `loom:blocked` | #152 |
| D | #154 | criterion throughput benchmark harness (A2C + PPO) | `loom:blocked` | #152 |

```
#151 ──> #152 ──┬──> #153  (example + learning curves + convergence test)
                └──> #154  (criterion throughput harness)
```

**Immediately buildable (no blockers):** #151 — labeled `loom:issue`. It is the only issue with no unmet dependency.

**Blocked until upstream merges:**
- #152 (`loom:blocked`) — flip to `loom:issue` once #151 lands on `main`.
- #153, #154 (`loom:blocked`) — independent of each other; flip both to `loom:issue` once #152 lands on `main`, at which point they can build in parallel.

#150 remains open as the tracking epic (keeps `loom:architect`).

**Comment by rjwalters** [comment]
## A2C epic complete ✅

All four child PRs implemented, reviewed, and merged to `main`:

| Issue | PR | Component |
|-------|-----|-----------|
| #151 | #155 | `A2cConfig` (builder + validate, classic defaults) |
| #152 | #156 | `A2cTrainer` + un-clipped A2C loss + fast smoke test |
| #153 | #158 | CartPole example + opt-in learning-curve CSV (A2C **and** PPO) + seeded convergence test |
| #154 | #157 | criterion throughput benchmark harness (`benches/trainer_throughput.rs`) |

### Benchmarks produced
- **Learning curves**: A2C solves CartPole (final avg episode length **500.00**, well above the 195 bar) in ~600k env steps; the convergence test passes in release (~17s). Both the A2C and PPO CartPole examples emit `(env_steps, mean_episode_reward)` CSV via the opt-in `CURVE_CSV` env var, seeded identically for same-env/same-seed overlay.
- **Throughput** (CPU ndarray, single update): A2C ~**264 K** elem/s vs PPO ~**241 K** elem/s on `train_step`; full rollout+update ~**184 K** vs ~**138 K** env-steps/s. A2C is consistently faster, as expected (single un-clipped update vs PPO's clipped surrogate + value clip).

A2C is now Thrust's fourth single-agent algorithm (PPO, DQN, SAC, A2C) and the first with a committed benchmark harness. Closing this tracking issue.
